

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>1. Getting started &mdash; coralME 1.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/jupyter-sphinx.css?v=572af1d6" />
      <link rel="stylesheet" type="text/css" href="_static/thebelab.css" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=56dcb7b8"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="_static/thebelab-helper.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2. Description of Inputs" href="BasicInputs.html" />
    <link rel="prev" title="Welcome to coralME documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            coralME
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">1. Getting started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Reconstruct-with-coralME">1.1. Reconstruct with coralME</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-packages">1.1.1. Import packages</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Define-path-to-inputs">1.1.2. Define path to inputs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Create-builder">1.1.3. Create builder</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Generate-files">1.1.4. Generate files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Build-ME-model">1.1.5. Build ME-model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Troubleshoot-ME-model">1.1.6. Troubleshoot ME-model</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Understanding-the-file-structure">1.2. Understanding the file structure</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Curate-manually">1.3. Curate manually</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BasicInputs.html">2. Description of Inputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="AddingBioCycDatabase.html">3. Adding annotation from a BioCyc Database</a></li>
<li class="toctree-l1"><a class="reference internal" href="coralMEArchitecture.html">4. Architecture of coralME</a></li>
<li class="toctree-l1"><a class="reference internal" href="FullReconstruction.html">5. Reconstructing with BioCyc information and curated files</a></li>
<li class="toctree-l1"><a class="reference internal" href="ReconstructionOSM.html">6. Reconstructing a ME-model from the OSM</a></li>
<li class="toctree-l1"><a class="reference internal" href="LoadingMEmodels.html">7. Loading and inspecting ME-models</a></li>
<li class="toctree-l1"><a class="reference internal" href="InspectingFluxes.html">8. Inspecting predicted fluxes</a></li>
<li class="toctree-l1"><a class="reference internal" href="EfficientSolving.html">9. Memory- and time-efficient solving of ME-models</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">10. Frequently Asked Questions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">coralME</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">1. </span>Getting started</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/GettingStarted.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Getting-started">
<h1><span class="section-number">1. </span>Getting started<a class="headerlink" href="#Getting-started" title="Link to this heading"></a></h1>
<section id="Reconstruct-with-coralME">
<h2><span class="section-number">1.1. </span>Reconstruct with coralME<a class="headerlink" href="#Reconstruct-with-coralME" title="Link to this heading"></a></h2>
<p>Here we show an example to reconstruct a dME-model of <em>B. subtilis</em></p>
<section id="Import-packages">
<h3><span class="section-number">1.1.1. </span>Import packages<a class="headerlink" href="#Import-packages" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">coralme.builder.main</span> <span class="kn">import</span> <span class="n">MEBuilder</span>
<span class="kn">from</span> <span class="nn">importlib.resources</span> <span class="kn">import</span> <span class="n">files</span>
</pre></div>
</div>
</div>
</section>
<section id="Define-path-to-inputs">
<h3><span class="section-number">1.1.2. </span>Define path to inputs<a class="headerlink" href="#Define-path-to-inputs" title="Link to this heading"></a></h3>
<p>Define the paths to inputs and desired outputs</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">input</span> <span class="o">=</span> <span class="p">{</span>
  <span class="c1"># Inputs</span>
  <span class="s2">&quot;m-model-path&quot;</span><span class="p">:</span> <span class="s2">&quot;./helper_files/inputs/m_model.json&quot;</span><span class="p">,</span> <span class="c1"># Path to model file</span>
  <span class="s2">&quot;genbank-path&quot;</span><span class="p">:</span> <span class="s2">&quot;./helper_files/inputs/genome.gb&quot;</span><span class="p">,</span> <span class="c1"># Path to genome genbank file</span>

  <span class="c1"># Outputs</span>
  <span class="s2">&quot;df_gene_cplxs_mods_rxns&quot;</span><span class="p">:</span> <span class="s2">&quot;./helper_files/building_data/OSM.xlsx&quot;</span><span class="p">,</span> <span class="c1"># Desired output path of OSM</span>
  <span class="s2">&quot;out_directory&quot;</span><span class="p">:</span> <span class="s2">&quot;./helper_files/&quot;</span><span class="p">,</span> <span class="c1"># Output directory</span>
  <span class="s2">&quot;log_directory&quot;</span><span class="p">:</span> <span class="s2">&quot;./helper_files/&quot;</span><span class="p">,</span> <span class="c1"># Log directory</span>
  <span class="s2">&quot;locus_tag&quot;</span><span class="p">:</span> <span class="s2">&quot;old_locus_tag&quot;</span><span class="p">,</span> <span class="c1"># What IDs were used in the M-model? e.g. locus_tag, old_locus_tag</span>
  <span class="s2">&quot;ME-Model-ID&quot;</span> <span class="p">:</span> <span class="s2">&quot;EXAMPLE-BACILLUS-ME&quot;</span> <span class="c1"># Name of the ME-model</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<p>Load organism setup. For now we can use the minimal setup with biological numbers from E. coli.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">organism</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">files</span><span class="p">(</span><span class="s2">&quot;coralme&quot;</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;iJL1678b&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;-ME/minimal-organism.json&quot;</span>
</pre></div>
</div>
</div>
</section>
<section id="Create-builder">
<h3><span class="section-number">1.1.3. </span>Create builder<a class="headerlink" href="#Create-builder" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">builder</span> <span class="o">=</span> <span class="n">MEBuilder</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">organism</span><span class="p">],</span> <span class="o">**</span><span class="nb">input</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Generate-files">
<h3><span class="section-number">1.1.4. </span>Generate files<a class="headerlink" href="#Generate-files" title="Link to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">builder</span><span class="o">.</span><span class="n">generate_files</span><span class="p">(</span><span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initiating file processing...
~ Processing files for EXAMPLE-BACILLUS-ME...
Set parameter Username
Academic license - for non-commercial use only - expires 2025-09-03
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Checking M-model metabolites...                                            : 100.0%|██████████|   990/  990 [00:00&lt;00:00]
Checking M-model genes...                                                  : 100.0%|██████████|   844/  844 [00:00&lt;00:00]
Checking M-model reactions...                                              : 100.0%|██████████|  1250/ 1250 [00:00&lt;00:00]
Syncing optional genes file...                                             : 0.0%|          |     0/    0 [00:00&lt;?]
Looking for duplicates within datasets...                                  : 100.0%|██████████|     5/    5 [00:00&lt;00:00]
Gathering ID occurrences across datasets...                                : 100.0%|██████████|  1250/ 1250 [00:00&lt;00:00]
Solving duplicates across datasets...                                      : 0.0%|          |     0/    0 [00:00&lt;?]
Pruning GenBank...                                                         : 100.0%|██████████|     1/    1 [00:01&lt;00:00]
Updating Genbank file with optional files...                               : 0.0%|          |     0/    0 [00:00&lt;?]
Syncing optional files with genbank contigs...                             : 100.0%|██████████|     1/    1 [00:08&lt;00:00]
Modifying metabolites with manual curation...                              : 0.0%|          |     0/    0 [00:00&lt;?]
Modifying metabolic reactions with manual curation...                      : 0.0%|          |     0/    0 [00:00&lt;?]
Adding manual curation of complexes...                                     : 0.0%|          |     0/    0 [00:00&lt;?]
Getting sigma factors...                                                   : 100.0%|██████████|    17/   17 [00:00&lt;00:00]
Getting generics from Genbank contigs...                                   : 100.0%|██████████|     1/    1 [00:00&lt;00:00]
Getting TU-gene associations from optional TUs file...                     : 0.0%|          |     0/    0 [00:00&lt;?]
Gathering ribosome stoichiometry...                                        : 100.0%|██████████|    63/   63 [00:00&lt;00:00]
Adding protein location...                                                 : 100.0%|██████████|  4238/ 4238 [00:00&lt;00:00]
Purging M-model genes...                                                   : 100.0%|██████████|   844/  844 [00:00&lt;00:00]
Getting enzyme-reaction associations...                                    : 100.0%|██████████|  1250/ 1250 [00:04&lt;00:00]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Reading EXAMPLE-BACILLUS-ME done.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Gathering M-model compartments...                                          : 100.0%|██████████|     2/    2 [00:00&lt;00:00]
Fixing compartments in M-model metabolites...                              : 100.0%|██████████|   990/  990 [00:00&lt;00:00]
Fixing missing names in M-model reactions...                               : 100.0%|██████████|  1250/ 1250 [00:00&lt;00:00]
Updating enzyme reaction association...                                    : 100.0%|██████████|   902/  902 [00:00&lt;00:00]
Getting tRNA to codon dictionary from NC_000964.3                          : 100.0%|██████████|  4449/ 4449 [00:03&lt;00:00]
Checking defined translocation pathways...                                 : 0.0%|          |     0/    0 [00:00&lt;?]
Getting reaction Keffs...                                                  : 100.0%|██████████|   902/  902 [00:00&lt;00:00]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Writting the Organism-Specific Matrix...
Organism-Specific Matrix saved to ./helper_files/building_data/OSM.xlsx file.
File processing done.
</pre></div></div>
</div>
</section>
<section id="Build-ME-model">
<h3><span class="section-number">1.1.5. </span>Build ME-model<a class="headerlink" href="#Build-ME-model" title="Link to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">builder</span><span class="o">.</span><span class="n">build_me_model</span><span class="p">(</span><span class="n">overwrite</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initiating ME-model reconstruction...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Adding biomass constraint(s) into the ME-model...                          : 100.0%|██████████|    11/   11 [00:00&lt;00:00]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Read LP format model from file /tmp/tmp6wx5hvsz.lp
Reading time = 0.00 seconds
: 990 rows, 2500 columns, 10478 nonzeros
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Read LP format model from file /tmp/tmp053a6_ir.lp
Reading time = 0.00 seconds
: 990 rows, 2496 columns, 10342 nonzeros
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Adding Metabolites from M-model into the ME-model...                       : 100.0%|██████████|   990/  990 [00:00&lt;00:00]
Adding Reactions from M-model into the ME-model...                         : 100.0%|██████████|  1248/ 1248 [00:00&lt;00:00]
Adding Transcriptional Units into the ME-model from user input...          : 0.0%|          |     0/    0 [00:00&lt;?]
Adding features from contig NC_000964.3 into the ME-model...               : 100.0%|██████████|  4449/ 4449 [00:08&lt;00:00]
Updating all TranslationReaction and TranscriptionReaction...              : 100.0%|██████████|  9232/ 9232 [00:21&lt;00:00]
Removing SubReactions from ComplexData...                                  : 100.0%|██████████|  4340/ 4340 [00:00&lt;00:00]
Adding ComplexFormation into the ME-model...                               : 100.0%|██████████|  4340/ 4340 [00:00&lt;00:00]
Adding Generic(s) into the ME-model...                                     : 100.0%|██████████|     5/    5 [00:00&lt;00:00]
Processing StoichiometricData in ME-model...                               : 100.0%|██████████|  1020/ 1020 [00:00&lt;00:00]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ME-model was saved in the ./helper_files/ directory as MEModel-step1-EXAMPLE-BACILLUS-ME.pkl
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Adding tRNA synthetase(s) information into the ME-model...                 : 100.0%|██████████|   306/  306 [00:00&lt;00:00]
Adding tRNA modification SubReactions...                                   : 0.0%|          |     0/    0 [00:00&lt;?]
Associating tRNA modification enzyme(s) to tRNA(s)...                      : 0.0%|          |     0/    0 [00:00&lt;?]
Adding SubReactions into TranslationReactions...                           : 100.0%|██████████|  4238/ 4238 [00:00&lt;00:00]
Adding RNA Polymerase(s) into the ME-model...                              : 100.0%|██████████|    17/   17 [00:00&lt;00:00]
Associating a RNA Polymerase to each Transcriptional Unit...               : 0.0%|          |     0/    0 [00:00&lt;?]
Processing ComplexData in ME-model...                                      : 100.0%|██████████|     1/    1 [00:00&lt;00:00]
Adding ComplexFormation into the ME-model...                               : 100.0%|██████████|  4365/ 4365 [00:00&lt;00:00]
Adding SubReactions into TranslationReactions...                           : 100.0%|██████████|  4238/ 4238 [00:00&lt;00:00]
Adding Transcription SubReactions...                                       : 100.0%|██████████|  4449/ 4449 [00:00&lt;00:00]
Processing StoichiometricData in SubReactionData...                        : 0.0%|          |     0/    0 [00:00&lt;?]
Adding reaction subsystems from M-model into the ME-model...               : 100.0%|██████████|  1248/ 1248 [00:00&lt;00:00]
Processing StoichiometricData in ME-model...                               : 100.0%|██████████|  1020/ 1020 [00:00&lt;00:00]
Updating ME-model Reactions...                                             : 39.4%|███▉      |  6325/16069 [00:51&lt;01:18]
</pre></div></div>
</div>
</section>
<section id="Troubleshoot-ME-model">
<h3><span class="section-number">1.1.6. </span>Troubleshoot ME-model<a class="headerlink" href="#Troubleshoot-ME-model" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">builder</span><span class="o">.</span><span class="n">troubleshoot</span><span class="p">(</span><span class="n">growth_key_and_value</span> <span class="o">=</span> <span class="p">{</span> <span class="n">builder</span><span class="o">.</span><span class="n">me_model</span><span class="o">.</span><span class="n">mu</span> <span class="p">:</span> <span class="mf">0.001</span> <span class="p">})</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Understanding-the-file-structure">
<h2><span class="section-number">1.2. </span>Understanding the file structure<a class="headerlink" href="#Understanding-the-file-structure" title="Link to this heading"></a></h2>
<p>For a detailed explanation of the inputs, see For more information about these files see <a class="reference internal" href="BasicInputs.html"><span class="doc">Description of inputs</span></a></p>
<p><img alt="Drawing" src="_images/folder.png" /></p>
</section>
<section id="Curate-manually">
<h2><span class="section-number">1.3. </span>Curate manually<a class="headerlink" href="#Curate-manually" title="Link to this heading"></a></h2>
<ol class="arabic simple">
<li><p><strong>Copy</strong> all of the generated <em>reference files</em> in building_data/reference_files and replace accordingly in building_data/</p></li>
<li><p><strong>Go one by one</strong> through the files in building_data/ curating as needed! Important flags are risen in curation_notes.json to further guide you through curation.</p></li>
<li><p>Everytime you make a change, <strong>run the model through the troubleshooter!</strong> It will show you remaining gaps to look at, and the new curation notes might show new warnings.</p></li>
<li><p><strong>Keep iterating!</strong> You will have finished when no gaps are present, and all remaining warnings in curation notes are irrelevant.</p></li>
</ol>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to coralME documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="BasicInputs.html" class="btn btn-neutral float-right" title="2. Description of Inputs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Juan D. Tibocha-Bonilla, Rodrigo Santibanez-Palominos.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>