Search.setIndex({"alltitles": {"1. M-model and Genbank are consistent, but they are not consistent with the BioCyc files.": [[1, "1.-M-model-and-Genbank-are-consistent,-but-they-are-not-consistent-with-the-BioCyc-files."]], "2. M-model and Genbank are not consistent": [[1, "2.-M-model-and-Genbank-are-not-consistent"]], "Architecture of coralME": [[4, null]], "Build ME-model": [[2, "Build-ME-model"]], "Configuration": [[0, "Configuration"]], "Configuration of parameters (organism.json)": [[0, "Configuration-of-parameters-(organism.json)"]], "Configuration of paths to files (inputs.json)": [[0, "Configuration-of-paths-to-files-(inputs.json)"]], "Content": [[5, "content"]], "Create builder": [[2, "Create-builder"]], "Curate manually": [[2, "Curate-manually"]], "Define inputs in input.json.": [[2, "Define-inputs-in-input.json."]], "Define organism and inputs": [[2, "Define-organism-and-inputs"]], "Define parameters in organism.json.": [[2, "Define-parameters-in-organism.json."]], "Description": [[0, "Description"], [0, "id1"], [0, "id2"], [0, "id4"], [0, "id6"], [0, "id8"], [0, "id10"], [0, "id12"], [0, "id14"], [0, "id16"], [5, "description"]], "Description of Inputs": [[0, null]], "Download RNAs.txt": [[2, "Download-RNAs.txt"]], "Download TUs.txt": [[2, "Download-TUs.txt"]], "Download files from BioCyc": [[2, "Download-files-from-BioCyc"]], "Download genes.txt and sequences.fasta": [[2, "Download-genes.txt-and-sequences.fasta"]], "Download proteins.txt": [[2, "Download-proteins.txt"]], "Download proteins.txt, RNAs.txt and TUs.txt": [[2, "Download-proteins.txt,-RNAs.txt-and-TUs.txt"]], "Frequently Asked Questions": [[1, null]], "Gene dictionary (genes.txt) [optional]": [[0, "Gene-dictionary-(genes.txt)-[optional]"]], "Gene sequences (sequences.fasta) [optional]": [[0, "Gene-sequences-(sequences.fasta)-[optional]"]], "Generate files": [[2, "Generate-files"]], "Genome (genome.gb)": [[0, "Genome-(genome.gb)"]], "Getting started": [[2, null]], "Go to Tools>Special SmartTables": [[2, "Go-to-Tools>Special-SmartTables"]], "Homology()": [[4, "Homology()"]], "How to curate?": [[3, "How-to-curate?"]], "How to manually curate a ME-model using coralME": [[3, null]], "Import packages": [[2, "Import-packages"]], "Indices and tables": [[5, "indices-and-tables"]], "Initialize the folder for your organism": [[2, "Initialize-the-folder-for-your-organism"]], "M-model (m_model.json)": [[0, "M-model-(m_model.json)"]], "MEBuilder()": [[4, "MEBuilder()"]], "MEReconstruction()": [[4, "MEReconstruction()"]], "Manual input files": [[3, "Manual-input-files"]], "My gene identifiers are not consistent, what should I do?": [[1, "My-gene-identifiers-are-not-consistent,-what-should-I-do?"]], "Optional": [[0, "Optional"]], "Organism()": [[4, "Organism()"]], "Proteins (proteins.txt) [optional]": [[0, "Proteins-(proteins.txt)-[optional]"]], "RNAs (RNAs.txt) [optional]": [[0, "RNAs-(RNAs.txt)-[optional]"]], "Reconstruct with coralME": [[2, "Reconstruct-with-coralME"]], "Required": [[0, "Required"]], "Requirements": [[0, "Requirements"], [0, "id3"], [0, "id5"], [0, "id7"], [0, "id9"], [0, "id11"], [0, "id13"], [0, "id15"], [0, "id17"]], "TUs (TUs.txt) [optional]": [[0, "TUs-(TUs.txt)-[optional]"]], "Troubleshoot ME-model": [[2, "Troubleshoot-ME-model"]], "Types of inputs": [[0, "Types-of-inputs"]], "Welcome to coralME documentation!": [[5, null]]}, "docnames": ["BasicInputs", "FAQ", "GettingStarted", "ManualCuration", "coralMEArchitecture", "index"], "envversion": {"nbsphinx": 4, "sphinx": 64, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.todo": 2, "sphinx.ext.viewcode": 1}, "filenames": ["BasicInputs.ipynb", "FAQ.ipynb", "GettingStarted.ipynb", "ManualCuration.ipynb", "coralMEArchitecture.ipynb", "index.rst"], "indexentries": {}, "objects": {"coralme.builder.curation": [[3, 0, 1, "", "AminoacidtRNASynthetase"], [3, 0, 1, "", "CleavedMethionine"], [3, 0, 1, "", "EnzymeReactionAssociation"], [3, 0, 1, "", "ExcisionMachinery"], [3, 0, 1, "", "FoldingDict"], [3, 0, 1, "", "GenericDict"], [3, 0, 1, "", "LipidModifications"], [3, 0, 1, "", "LipoproteinPrecursors"], [3, 0, 1, "", "MEMetabolites"], [3, 0, 1, "", "ManualComplexes"], [3, 0, 1, "", "OrphanSpontReactions"], [3, 0, 1, "", "PeptideReleaseFactors"], [3, 0, 1, "", "ProteinLocation"], [3, 0, 1, "", "RNADegradosome"], [3, 0, 1, "", "RNAModificationMachinery"], [3, 0, 1, "", "RNAModificationTargets"], [3, 0, 1, "", "ReactionCorrections"], [3, 0, 1, "", "ReactionMatrix"], [3, 0, 1, "", "RhoIndependent"], [3, 0, 1, "", "RibosomeStoich"], [3, 0, 1, "", "RibosomeSubreactions"], [3, 0, 1, "", "Sigmas"], [3, 0, 1, "", "SpecialModifications"], [3, 0, 1, "", "SpecialtRNASubreactions"], [3, 0, 1, "", "SubreactionMatrix"], [3, 0, 1, "", "SubsystemClassification"], [3, 0, 1, "", "TerminationSubreactions"], [3, 0, 1, "", "TranscriptionSubreactions"], [3, 0, 1, "", "TranslocationMultipliers"], [3, 0, 1, "", "TranslocationPathways"]], "coralme.builder.curation.ManualComplexes": [[3, 1, 1, "", "org_data"]]}, "objnames": {"0": ["py", "class", "Python class"], "1": ["py", "property", "Python property"]}, "objtypes": {"0": "py:class", "1": "py:property"}, "terms": {"": [1, 2, 3, 5], "0": [0, 2, 3], "001": 2, "1": [0, 3], "10": [], "1341": 3, "16": 3, "17": 3, "1e": [], "2": [0, 3], "20": 3, "20a": 3, "21": 3, "25": 0, "2fe2": 3, "3": 0, "30": 3, "32": 3, "4": [0, 3, 4], "5": [0, 2], "50": 3, "7": 0, "7617": 3, "7824": 3, "8": [], "AND": 3, "But": 2, "For": 2, "If": [0, 1, 3, 4], "It": [0, 1, 3, 4], "No": 3, "OR": 3, "One": 1, "That": 1, "The": [0, 2, 4, 5], "These": 3, "To": 2, "__": 3, "__coralm": 3, "_refer": 3, "about": [2, 4, 5], "absolut": [], "access": [0, 1, 4], "accoa_c": 3, "accordingli": [1, 3], "acid": [0, 3], "acra": 3, "across": 1, "act": 4, "ad": [1, 2, 3], "adapt": 4, "add": [2, 3], "adnt2pp": 3, "advis": 4, "after": [2, 3], "ala__l_c": 3, "ala_rs_tetra_mod_zn2": 3, "algorithm": [3, 5], "all": [0, 1, 2, 3, 4], "almost": 4, "also": 3, "although": 4, "alwai": 3, "amino": [0, 3], "amino_acid": 3, "amino_acid_trna_synthetas": 3, "aminoacidtrnasynthetas": 3, "an": [0, 1, 2, 3, 4, 5], "ani": [3, 4, 5], "annot": 0, "appear": [1, 2], "appli": 2, "applic": [2, 4], "approach": 1, "ar": [0, 2, 3, 4, 5], "architectur": [2, 5], "arg": 4, "argument": 4, "ask": 5, "assembli": 1, "associ": [0, 3], "assum": 1, "attribut": 4, "autom": 4, "automat": [3, 5], "avail": [2, 4, 5], "awai": 2, "b": [2, 4], "b0014": 3, "b0202": 3, "b0344": 3, "b0463": 3, "b1855": 3, "b3461": 3, "b4154": 3, "back": 3, "bacteria": 4, "bar": [1, 2], "base": 4, "basic": 5, "befor": [0, 3], "best": [1, 4], "between": [1, 3, 4], "biocyc": [0, 5], "bioinformat": 1, "biomass": 0, "biopython": 0, "blast": 4, "bnum": 3, "bool": 4, "bsu02690": 3, "bsu16650": 3, "bsu24740": 3, "bsu27650": 3, "bsu32670": 3, "bsu32700": 3, "bsu32710": 3, "bsu35300": 3, "build": [3, 5], "build_me_model": [2, 4], "builder": [3, 4], "building_data": 3, "calcul": [], "call": 4, "can": [0, 1, 2, 3, 4], "case": [0, 1], "central_afn": 3, "central_c": 3, "chang": [1, 3], "chdir": [], "check": 2, "choos": 4, "class": [3, 4], "classif": 3, "classifi": [1, 3], "cleavag": 3, "cleaved_methionin": 3, "cleaved_methionine_gen": 3, "cleavedmethionin": 3, "click": 0, "cluster": 3, "co": 0, "coa_c": 3, "cobalt": 3, "cobalt2_": 3, "cobalt2_c": 3, "cobalt2tpp": 3, "cobram": 5, "cobrapi": 0, "code": [3, 4], "codh_fe_load": 3, "codon": [], "coli": 4, "column": [0, 1, 2, 3], "common": [0, 2], "compart": 3, "complement": [0, 4, 5], "complet": 5, "complex": [0, 2, 3], "complex_compart": 3, "complex_id": 3, "compliant": 0, "compon": [0, 3], "compos": 4, "composit": [0, 3], "comprehens": 5, "config": [3, 4], "configur": [2, 4], "consid": [1, 4], "consist": [0, 2, 5], "constraint": 0, "contain": [0, 1, 3, 4], "contradictori": 5, "convent": 1, "convert": 1, "coordin": 4, "copi": [1, 2, 3], "coralm": 0, "correct": [0, 1, 3], "correctli": 0, "correspond": 1, "could": [1, 2], "cplex": [], "cplx0": 3, "cplx_dummi": 3, "creat": 3, "critic": 1, "cs_c": 3, "cs_cyto_import": 3, "cs_p": 3, "curat": [4, 5], "curation_not": 3, "custom": 0, "cutoff": 4, "d": 3, "databas": [0, 1, 4], "dataset": 3, "default": [0, 2], "defin": [0, 3], "definit": 3, "degradosom": 3, "demonstr": 5, "depend": 1, "describ": 5, "descript": [2, 3], "detail": 2, "dict": 4, "dictionari": [1, 3, 4], "didn": 1, "differ": 1, "dir": 2, "direct": [0, 2], "directli": [0, 3], "dme": 2, "do": [2, 5], "doe": 3, "download": [0, 1, 5], "dropdown": 2, "dure": [], "dusb_mono": 3, "e": [0, 1, 3, 4, 5], "ecolim": [4, 5], "edit": [1, 2], "editor": 1, "either": 3, "elong": 3, "elongation_subreact": 3, "enabl": 0, "end": 0, "eno_dim_mod_mg2": 3, "ensur": [1, 2], "entir": [0, 4], "entri": 5, "environment": 0, "enz_rxn_assoc_df": 3, "enzym": [3, 4], "enzyme_reaction_associ": 3, "enzymereactionassoci": 3, "estim": 3, "etc": 0, "evalu": 4, "even": 1, "everytim": 3, "exampl": [0, 2, 3], "exchang": 4, "excis": 3, "excision_machineri": 3, "excisionmachineri": 3, "exist": [0, 5], "export": 0, "factor": 3, "fals": [2, 3, 4], "fe": 3, "feasibl": 2, "feel": 2, "fes_transf": 3, "field": 0, "file": [4, 5], "files_": 3, "fill": 3, "final": 3, "find": 0, "finish": 3, "first": 3, "fix": [], "flag": 3, "float": 4, "flux": 3, "fold": 3, "folder": [4, 5], "folding_dict": 3, "foldingdict": 3, "follow": [1, 3], "format": 3, "formula": 3, "found": 1, "four": 5, "frame": 0, "free": 2, "frequent": 5, "from": [0, 1, 3, 4, 5], "function": 5, "further": [3, 4], "g": [0, 1, 4], "gap": [2, 3], "gapfil": [], "gapfill_cofactor": [], "genbank": [0, 2], "gene": [3, 5], "gene_reaction_rul": 3, "gener": [1, 3, 4, 5], "generate_fil": 2, "generic_16sm4cm1402": 3, "generic_23s_rrna": 3, "generic_5s_rrna": 3, "generic_compon": 3, "generic_dict": 3, "generic_rf": 3, "genericdict": 3, "genom": 1, "get": [1, 5], "get_homologi": [], "get_trna_to_codon": [], "gettingstart": 3, "give": 2, "go": 3, "gpr": 0, "gram": 4, "grea_mono": 3, "greb_mono": 3, "groel_dependent_fold": 3, "growth_key_and_valu": 2, "gtp_bound_30s_assembly_factor_phase1": 3, "guess": [], "guid": [3, 5], "gurobi": [], "h": 3, "h_c": 3, "h_p": 3, "ha": [0, 4, 5], "happen": 1, "hardest": 1, "have": [1, 2, 3], "help": 1, "helper_fil": 2, "here": [2, 3, 4], "high": 2, "hit": 4, "homolog": 4, "homologi": 5, "how": [2, 5], "howev": 1, "i": [0, 2, 4, 5], "id": [0, 1, 2, 3], "identifi": [0, 4, 5], "implement": 5, "import": 3, "inconsist": 1, "independ": 3, "index": [0, 2, 5], "infer": 4, "inform": [0, 1, 2, 3, 4, 5], "initi": [3, 5], "initiation_subreact": 3, "input": [4, 5], "instanc": 3, "integr": 5, "intend": 5, "intermedi": 3, "intern": 0, "ipynb": 3, "irrelev": 3, "is_refer": 4, "is_revers": 3, "is_spontan": 3, "isca_tetra": 3, "issu": 1, "iter": 3, "its": 4, "json": [3, 4], "keep": 3, "keff": 3, "keyword": 4, "know": 1, "known": [], "kwarg": 4, "last": 1, "latter": 0, "left": 0, "less": 4, "lgt_monom": 3, "ligas": 3, "lipid": 3, "lipid_mod": 3, "lipid_modif": 3, "lipidmodif": 3, "lipoprotein": 3, "lipoprotein_precursor": 3, "lipoproteinprecursor": 3, "list": [1, 2], "load": [2, 3], "locat": [0, 2, 3], "locu": 0, "locus_tag": [0, 1], "look": [1, 3], "low": 2, "lspa_monom": 3, "m": [2, 3], "machineri": 3, "main": [1, 2, 4, 5], "make": [0, 1, 3, 4], "mani": [2, 3], "manipul": 4, "manual": [4, 5], "manual_complex": 3, "manualcomplex": 3, "map": 3, "mark": 3, "matrix": [0, 3, 5], "mayb": 1, "me": [0, 4, 5], "me_id": 3, "me_met": 3, "me_metabolit": 3, "me_model": 2, "meant": [0, 3], "mebuild": [2, 5], "mechan": 3, "median": 3, "memetabolit": 3, "meprocess": 4, "mereconstruct": 5, "met": 0, "met_typ": [], "metabol": 0, "metabolit": 3, "methionin": 3, "method": [3, 4], "metroubleshoot": 4, "mfd_mono_mod_mg2": 3, "microb": 1, "might": [2, 3], "misc": [0, 2], "miss": 0, "mistak": 1, "mod": 3, "mod_acetyl_c": 3, "model": [4, 5], "modif": 3, "modifi": [1, 2, 3], "modul": 5, "monom": 3, "moreov": 4, "most": 3, "mu": 2, "much": 1, "multipli": 3, "must": [0, 1, 2], "my": 5, "n": 3, "name": [0, 2, 3, 4], "ncbi": 0, "necessari": [4, 5], "need": [2, 3], "neg": 4, "network": 0, "new": [1, 3], "none": [], "note": [0, 2, 3, 5], "nucleotid": 0, "nupc": 3, "nupg": 3, "nusa_mono": 3, "nusg_mono": 3, "o": [], "object": [3, 4, 5], "often": 1, "old_locus_tag": 0, "onc": 2, "one": [1, 2, 3], "ones": 3, "onli": 1, "operon": 0, "option": [1, 2], "org": [3, 4], "org_data": 3, "organ": [3, 5], "origin": 4, "orphan": 3, "orphan_and_spont_react": 3, "orphanspontreact": 3, "osm": [], "other": [3, 4], "overal": 1, "overwrit": [2, 3], "packag": 5, "page": 5, "paramet": [3, 4], "parent": 4, "pass": 4, "past": 1, "path": 4, "pathwai": 3, "peptid": 3, "peptide_compartment_and_pathwai": 3, "peptide_release_factor": 3, "peptidereleasefactor": 3, "perform": [3, 5], "permeas": 3, "pg_pe_160": 3, "pickl": [], "pipelin": 5, "plasma_membran": 3, "platform": 1, "polymeras": 3, "polypeptid": [0, 2], "pore": 3, "posit": [0, 3, 4], "possibl": 1, "post_transcriptional_modification_of_rna": 3, "pp0001": 1, "pp_0001": 1, "pre": 3, "precursor": 3, "predict": 4, "prepar": [], "prepare_model": [], "present": 3, "prfa_mono": 3, "prfa_mono_mediated_termin": 3, "probabl": 1, "problem": 1, "process": [2, 3, 4, 5], "product": 0, "prokaryot": 5, "properti": [2, 3], "protein": 3, "protein_compart": 3, "protein_correct": 3, "protein_loc": 3, "proteinloc": 3, "provid": [0, 4], "put": 2, "py": 4, "question": 5, "quick": 1, "quickest": 2, "reaction": [0, 3], "reaction_correct": 3, "reaction_id": 3, "reaction_matrix": 3, "reaction_median_keff": 3, "reactioncorrect": 3, "reactionmatrix": 3, "read": [0, 3, 4], "readili": 5, "reciproc": 4, "reconstruct": [0, 1, 4, 5], "ref": 4, "refer": [3, 4], "reference_fil": 3, "regard": 0, "rel": [], "releas": 3, "release_factor": 3, "remain": 3, "remov": [3, 5], "replac": 3, "requir": [3, 5], "resort": 1, "resourc": 5, "respect": 4, "result": 4, "retriev": 4, "rho": 3, "rho_independ": 3, "rhoindepend": 3, "ribosom": 3, "ribosomal_protein": 3, "ribosome_stoich": 3, "ribosome_subreact": 3, "ribosomestoich": 3, "ribosomesubreact": 3, "right": [0, 2], "risen": 3, "rna": 3, "rna_degradosom": 3, "rna_modif": 3, "rna_modification_df": 3, "rna_modification_target": 3, "rnadegradosom": 3, "rnamodificationmachineri": 3, "rnamodificationtarget": 3, "rnap_32h": 3, "rnase_e_tetra_mod_zn2": 3, "role": 4, "rpla_mono": 3, "rpoh_mono": 3, "rpoz_mono_mod_mg2": 3, "rpsd_mono": 3, "rrna": [0, 2, 3], "rrna_contain": 3, "rsmh_mono": 3, "rsmi_mono": 3, "run": 3, "s_amino_acids_and_related_molecul": 3, "same": [0, 1, 2, 4], "sasa": 3, "save": 3, "savefil": [], "scaffold": 3, "scenario": 1, "search": [1, 2, 5], "sec": 3, "secondari": 3, "secyeg": 3, "see": [0, 1, 2], "selenocystein": 3, "separ": 0, "ser": 3, "set": [2, 3, 4], "should": [0, 2, 5], "show": [0, 2, 3], "sigma": 3, "sigma_factor": 3, "signific": 2, "sinc": 1, "skip": [], "smarttabl": [0, 1], "so": 1, "solv": 1, "solvem": 5, "solver": [], "some": 2, "sometim": 2, "somewher": 1, "sourc": [1, 3, 4], "special": [0, 3], "special_modif": 3, "special_trna_subreact": 3, "specialmodif": 3, "specialtrnasubreact": 3, "specif": [4, 5], "specifi": 3, "spontan": 3, "spreadsheet": 0, "stage": [2, 3], "standard": [0, 2], "start": [1, 5], "step": 5, "still": 1, "stoich": 3, "stoichiometri": 3, "store": [3, 4], "str": 4, "strain": 2, "subreact": 3, "subreaction_matrix": 3, "subreactionmatrix": 3, "substitut": 3, "subsystem": [0, 3], "subsystem_classif": 3, "subsystemclassif": 3, "subtili": [2, 4], "subunit": 3, "sufbc2d": 3, "sufbcd": 3, "sufbcd_c": 3, "sufbcd_mod_2fe3": 3, "sure": [0, 1], "symbol": [], "sympi": [], "synchron": 5, "synonym": 1, "synthesi": 3, "synthetas": 3, "t": 1, "tab": 0, "tabl": 0, "tag": 0, "take": 0, "target": 3, "tata_monom": 3, "tate_monom": 3, "templat": 5, "termin": 3, "termination_subreact": 3, "terminationsubreact": 3, "text": 1, "than": 4, "thei": 3, "them": [0, 2, 3], "thi": [0, 1, 2, 3, 4, 5], "three": 1, "through": 3, "thu": 1, "time": 3, "too": 2, "total": 0, "transcrib": 0, "transcript": [0, 2, 3], "transcription_normal_rho_independ": 3, "transcription_subreact": 3, "transcriptionsubreact": 3, "transform": 2, "translat": 3, "transloc": 3, "translocase_pathwai": 3, "translocation_multipli": 3, "translocation_pathwai": 3, "translocationmultipli": 3, "translocationpathwai": 3, "transport": 3, "tri": 1, "trna": [0, 2, 3], "troubleshoot": 3, "true": [2, 3, 4], "try": 1, "tus_from_biocyc": 3, "tutori": 2, "two": 4, "txt": [1, 3], "type": [3, 5], "uaa": 3, "umbrella": 3, "under": 0, "undergo": 3, "underscor": 1, "unit": [0, 2, 3], "updat": 4, "us": [0, 1, 2, 4, 5], "user": [4, 5], "valu": [2, 4], "verbos": 4, "via": 3, "view": 0, "wa": [0, 1, 4], "wai": [1, 2], "warn": 3, "we": [1, 2, 4], "webpag": 2, "well": 3, "were": 4, "what": 5, "when": [1, 3], "which": [1, 3], "while": 3, "win32": [], "within": 2, "work": 5, "wrong": 1, "xml": 0, "yet": 3, "yidc_monom": 3, "yield": 1, "you": [0, 1, 2, 3], "your": [1, 3, 5]}, "titles": ["<span class=\"section-number\">2. </span>Description of Inputs", "<span class=\"section-number\">5. </span>Frequently Asked Questions", "<span class=\"section-number\">1. </span>Getting started", "<span class=\"section-number\">4. </span>How to manually curate a ME-model using coralME", "<span class=\"section-number\">3. </span>Architecture of coralME", "Welcome to coralME documentation!"], "titleterms": {"1": 1, "2": 1, "ar": 1, "architectur": 4, "ask": 1, "biocyc": [1, 2], "build": 2, "builder": 2, "configur": 0, "consist": 1, "content": 5, "coralm": [2, 3, 4, 5], "creat": 2, "curat": [2, 3], "defin": 2, "descript": [0, 5], "dictionari": 0, "do": 1, "document": 5, "download": 2, "fasta": [0, 2], "file": [0, 1, 2, 3], "folder": 2, "frequent": 1, "from": 2, "gb": 0, "genbank": 1, "gene": [0, 1, 2], "gener": 2, "genom": 0, "get": 2, "go": 2, "homologi": 4, "how": 3, "i": 1, "identifi": 1, "import": 2, "indic": 5, "initi": 2, "input": [0, 2, 3], "json": [0, 2], "m": [0, 1], "m_model": 0, "manual": [2, 3], "me": [2, 3], "mebuild": 4, "mereconstruct": 4, "model": [0, 1, 2, 3], "my": 1, "option": 0, "organ": [0, 2, 4], "packag": 2, "paramet": [0, 2], "path": 0, "protein": [0, 2], "question": 1, "reconstruct": 2, "requir": 0, "rna": [0, 2], "sequenc": [0, 2], "should": 1, "smarttabl": 2, "special": 2, "start": 2, "tabl": 5, "thei": 1, "tool": 2, "troubleshoot": 2, "tu": [0, 2], "txt": [0, 2], "type": 0, "us": 3, "welcom": 5, "what": 1, "your": 2}})